src/cpp/shared/town/town.cpp

```diff
@@ -1146,46 +1145,50 @@ void townManager::SetupCastle(heroWindow *window, int a3) {
   tag_message evt;
   casWin = window;
 
-  for(int i = 0; i < 18; ++i) {
+  for(int i = 0; i < NUM_NON_DWELLING_BUILDINGS; ++i) {
     castleSlotsUse[i] = castleSlotsBase[i];
     if((signed int)castleSlotsBase[i] >= 20
       && (signed int)castleSlotsBase[i] <= 24
       && ((1 << castleSlotsBase[i]) & this->castle->buildingsBuiltFlags
         || (1 << (castleSlotsBase[i] + 5)) & this->castle->buildingsBuiltFlags
-        || castleSlotsBase[i] == 24 && this->castle->factionID == 3 && (*((unsigned char*)&(this->castle->buildingsBuiltFlags)+3)) & 0x40)
+        || castleSlotsBase[i] == 24 && this->castle->factionID == FACTION_WARLOCK && (*((unsigned char*)&(this->castle->buildingsBuiltFlags)+3)) & 0x40)
       && (1 << (castleSlotsBase[i] + 5)) & gTownEligibleBuildMask[this->castle->factionID]) {
       if(castleSlotsBase[i] == 24
-        && this->castle->factionID == 3
+        && this->castle->factionID == FACTION_WARLOCK
         && ((*((unsigned char*)&(this->castle->buildingsBuiltFlags)+3)) & 0x20 || (*((unsigned char*)&(this->castle->buildingsBuiltFlags)+3)) & 0x40))
```

Comment:
This is: this->castle->buildingsBuiltFlags & (1 << BUILDING_UPGRADE_5). The other bit tests around here can be refactored similarly.

Reply:
Can you please describe how you shortened it to this form? I am not very good at bit shifting, would be very helpful for future!

Reply:
Here are the steps to figure this out. Make sure you understand each step, and tell me if you don't understand one:

1)

(*((unsigned char*)&(this->castle->buildingsBuiltFlags)+3)

This refers to bytes 24-31 of this->castle->buildingsBuiltFlags.

2) 

0x20 is equal to (1 << 5).

The "& 0x20" is testing the 6th byte (byte index 5) of *((unsigned char*)&(this->castle->buildingsBuiltFlags)+3), which is byte 29 of this->castle->buildingsBuiltFlags . 

3) 

So, that expression is equivalent to (this->castle->buildingsBuiltFlags & (1 << 29)). And 29 == BUILDING_UPGRADE_5 .

Reply:
Done. Thank you!